[k8s] Kubernetes Docs (#2324)
skypilot-org/skypilot: docs/source/reference/kubernetes/kubernetes-setup.rst

Reviewer: Nit: setup is a noun; used "set up" throughout?

.. _kubernetes-setup:

Kubernetes Cluster Setup
========================

.. note::
    This is a guide for cluster administrators on how to setup Kubernetes clusters
    for use with SkyPilot.

    If you are a SkyPilot user and your cluster administrator has already setup a cluster
    and shared a kubeconfig file with you, :ref:`Submitting tasks to Kubernetes <kubernetes-instructions>`
    explains how to submit tasks to your cluster.

> SkyPilot on Kubernetes is designed to work with most Kubernetes distributions and deployment environments.

Reviewer: SkyPilot's Kubernetes support?

To connect to a Kubernetes cluster, SkyPilot needs:

* An existing Kubernetes cluster running Kubernetes v1.20 or later.
* A `Kubeconfig <kubeconfig>`_ file containing access credentials and namespace to be used.

Deployment Guides
-----------------
Below we show minimal examples to setup a new Kubernetes cluster in different environments, including hosted services on the cloud, and generating kubeconfig files which can be :ref:`used by SkyPilot <kubernetes-instructions>`.

..

>   TODO(romilb) - Add a table of contents/grid cards for each deployment environment.

Reviewer: +1, or how about adding just an item list now would be great

.. _kubernetes-setup-kind:

Deploying locally on your Laptop
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> If you want to try out SkyPilot on Kubernetes on your laptop or run SkyPilot

Reviewer: ```suggestion
To try out SkyPilot on Kubernetes on your laptop or run SkyPilot
```

"To xxx, yyy" has been used throughout our docs. Stay consistent?

tasks locally without requiring any cloud access, we provide the
:code:`sky local up` CLI to create a 1-node Kubernetes cluster locally.

Under the hood, :code:`sky local up` uses `kind <https://kind.sigs.k8s.io/>`_,
a tool for creating a Kubernetes cluster on your local machine.
It runs a Kubernetes cluster inside a container, so no setup is required.

1. Install `Docker <https://docs.docker.com/engine/install/>`_ and `kind <https://kind.sigs.k8s.io/>`_.
2. Run :code:`sky local up` to launch a Kubernetes cluster and automatically configure your kubeconfig file:

>     .. code-block:: console

Reviewer: Hmm, on my local these codeblocks rendered in blockquotes:

<img width="821" alt="Screen Shot 2023-09-11 at 22 22 46" src="https://github.com/skypilot-org/skypilot/assets/592670/91be6c25-16eb-4732-a5a7-9620475e524d">

Author: Ah good catch - I've fixed this now:
<img width="859" alt="image" src="https://github.com/skypilot-org/skypilot/assets/4416605/b681853e-b7a7-4d88-982e-4d5bf9fa1671">

        $ sky local up

3. Run :code:`sky check` and verify that Kubernetes is enabled in SkyPilot. You can now run SkyPilot tasks on this locally hosted Kubernetes cluster using :code:`sky launch`.
4. After you are done using the cluster, you can remove it with :code:`sky local down`. This will terminate the KinD container and switch your kubeconfig back to it's original context:

    .. code-block:: console

        $ sky local down

.. note::
    We recommend allocating at least 4 or more CPUs to your docker runtime to
    ensure kind has enough resources. See instructions
    `here <https://docs.docker.com/desktop/settings/linux/>`_.

.. note::

>     KinD does not support multiple nodes and GPUs.

Reviewer: Be consistent, kind vs KinD

    It is not recommended for use in a production environment.

>     If you want to run a private On-Prem cluster, see the section on `On-Prem deployment <Deploying on On-Prem Clusters>`_ for more.

Reviewer: fix link

.. _kubernetes-setup-gke:

> Deploying on GKE

Reviewer: "GKE" vs "AWS EKS" not consistent. Either add GCP to former or drop AWS from latter?

^^^^^^^^^^^^^^^^

1. Create a GKE standard cluster with at least 1 node. We recommend creating nodes with at least 4 vCPUs.

> 2. Get the kubeconfig for your cluster. This will automatically update ``~/.kube/config`` with new kubecontext for the GKE cluster:

Reviewer: ```suggestion
2. Get the kubeconfig for your cluster. The following commands will automatically update ``~/.kube/config`` with new kubecontext for the GKE cluster:
```

    .. code-block:: console

        $ gcloud container clusters get-credentials <cluster-name> --region <region>

        # Example:
        # gcloud container clusters get-credentials testcluster --region us-central1-c

3. [If using GPUs] If your GKE nodes have GPUs, you may need to to
   `manually install <https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/create-cluster-kubeadm/>`_
   nvidia drivers. You can do so by deploying the daemonset
   depending on the OS of your nodes:

    .. code-block:: console

        # For Container Optimized OS (COS) based nodes:
        $ kubectl apply -f https://raw.githubusercontent.com/GoogleCloudPlatform/container-engine-accelerators/master/nvidia-driver-installer/cos/daemonset-preloaded.yaml

>         # For Ubuntu (COS) based nodes:

Reviewer: COS in both cases?

    .. code-block:: console

        $ sky check

.. note::
    GKE autopilot clusters are currently not supported. Only GKE standard clusters are supported.

.. _kubernetes-setup-eks:

Deploying on AWS EKS
^^^^^^^^^^^^^^^^^^^^

1. Create a EKS cluster with at least 1 node. We recommend creating nodes with at least 4 vCPUs.

2. TODO - Add instructions on getting kubeconfig.

3. [If using GPUs] EKS clusters already come with Nvidia drivers setup. However, you will need to label the nodes with the GPU type. Use the SkyPilot node labelling tool to do so:

    .. code-block:: console

        python -m sky.utils.kubernetes.gpu_labeler

   This will create a job on each node to read the GPU type from `nvidia-smi` and assign a ``skypilot.co/accelerator`` label to the node. You can check the status of these jobs by running:

    .. code-block:: console

        kubectl get jobs -n kube-system

4. Verify your kubeconfig (and GPU support, if available) is correctly setup by running :code:`sky check`:

    .. code-block:: console

        $ sky check

.. _kubernetes-setup-onprem:

Deploying on On-Prem Clusters
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> You can also deploy Kubernetes on your On-Prem clusters using off-the-shelf tools,

Reviewer: ```suggestion
You can also deploy Kubernetes on your on-prem clusters using off-the-shelf tools,
```
nit: on-prem throughout?

such as `kubeadm <https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/create-cluster-kubeadm/>`_,
`k3s <https://docs.k3s.io/quick-start>`_ or
`Rancher <https://ranchermanager.docs.rancher.com/v2.5/pages-for-subheaders/kubernetes-clusters-in-rancher-setup>`_.
Please follow their respective guides to deploy your Kubernetes cluster.

Setting up GPU support
~~~~~~~~~~~~~~~~~~~~~~
If your Kubernetes cluster has Nvidia GPUs, make sure you have the Nvidia
device plugin installed (i.e., ``nvidia.com/gpu`` resource is available on each node).
Additionally, you will need to label each node in your cluster with the GPU type.
For example, a node with v100 GPUs must have a label :code:`skypilot.co/accelerators: v100`.

> We provide a convenience script that automatically detects GPU type and labels each node. You can run it with:

Reviewer: ```suggestion
We provide a convenience script that automatically detects GPU types and labels each node. You can run it with:
```

.. code-block:: console

 $ python -m sky.utils.kubernetes.gpu_labeler

 Created GPU labeler job for node ip-192-168-54-76.us-west-2.compute.internal
 Created GPU labeler job for node ip-192-168-93-215.us-west-2.compute.internal
 GPU labeling started - this may take a few minutes to complete.
 To check the status of GPU labeling jobs, run `kubectl get jobs --namespace=kube-system -l job=sky-gpu-labeler`
 You can check if nodes have been labeled by running `kubectl describe nodes` and looking for labels of the format `skypilot.co/accelerators: <gpu_name>`.

.. note::
 GPU labelling is not required on GKE clusters - SkyPilot will automatically use GKE provided labels. However, you will still need to install `drivers <https://cloud.google.com/kubernetes-engine/docs/how-to/gpus#installing_drivers>`_.

.. note::

>  To cleanup any leftover jobs from the GPU labelling process, run ``python -m sky.utils.kubernetes.gpu_labeler --cleanup``.

Reviewer: At this point, wondering what are these leftover jobs.

Author: Thanks for catching - reworded now

Deployment Guides
-----------------
Below we show minimal examples to set up a new Kubernetes cluster in different environments, including hosted services on the cloud, and generating kubeconfig files which can be :ref:`used by SkyPilot <kubernetes-instructions>`.

..
  TODO(romilb) - Add a table of contents/grid cards for each deployment environment.

* :ref:`Deploying locally on your laptop <kubernetes-setup-kind>`
* :ref:`Deploying on Google Cloud GKE <kubernetes-setup-gke>`
* :ref:`Deploying on Amazon EKS <kubernetes-setup-eks>`
* :ref:`Deploying on on-prem clusters <kubernetes-setup-onprem>`

.. _kubernetes-setup-kind:

Deploying locally on your laptop
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

To try out SkyPilot on Kubernetes on your laptop or run SkyPilot
tasks locally without requiring any cloud access, we provide the
:code:`sky local up` CLI to create a 1-node Kubernetes cluster locally.

Under the hood, :code:`sky local up` uses `kind <https://kind.sigs.k8s.io/>`_,
a tool for creating a Kubernetes cluster on your local machine.

> It runs a Kubernetes cluster inside a container, so no set up is required.

Reviewer: ```suggestion
It runs a Kubernetes cluster inside a container, so no setup is required.
```
